### PR TITLE
Apply --all if workspace is virtual

### PR DESCRIPTION
**Pull request by rsertelon**

### CODE REVIEW

**Comment by alexcrichton** [comment]
Awesome, thanks! Could you also add some tests for this as well?

**Comment by rsertelon** [comment]
Will do this evening (GMT+1/2)

**Comment by rsertelon** [comment]
I've fixed one test that failed because the virtual build was supposed to fail before this PR.

However, I don't really see a way of testing the modification appart from trying to run cargo <cmd impacted by changes> with a virtual build and verify it returns with a status 0.

Is that actually all that's testable?

**Comment by matklad** [comment]
>However, I don't really see a way of testing the modification appart from trying to run cargo with a virtual build and verify it returns with a status 0.

`cargo test` would be a convenient command to check that everything is tested: you'll need a workspace with a virtual manifest and two members, `foo` and `bar`. You then run `cargo test` in the virtual package and verify that stdout contains lines `test_foo`, `test_bar`.

**Comment by rsertelon** [comment]
Thanks for the pointer @matklad Won't have time this week-end, but will be back on it next week!

**Comment by rsertelon** [comment]
I think I've got this. I copied existing tests for virtual workspace, and removed the `--all` argument and if test fails, it means that the automatic add of `--all` is broken.

What do you think?

File: src/bin/bench.rs [hunk 1]
```diff
@@ -84,19 +86,24 @@ Compilation can be customized with the `bench` profile in the manifest.
...
+                     options.flag_quiet,
+                     &options.flag_color,
+                     options.flag_frozen,
+                     options.flag_locked)?;
+
     let root = find_root_manifest_for_wd(options.flag_manifest_path, config.cwd())?;
+    let ws = Workspace::new(&root, config)?;
 
-    let spec = if options.flag_all {
+    let spec = if options.flag_all || ws.is_virtual() {
```
Comment (matklad) [inline src/bin/bench.rs]: Hm, won't this break a use case when you are in the root of the workspace, but want to exercise a particular package?

I am thinking about something like this:

```
ws_root/
  Cargo.toml
  foo/
    Cargo.toml
  bar/
    Cargo.toml
```

```
$ pwd
~/projects/ws_root/
$ cargo test --pkg foo
```

If I am right, and the tests are green, this also means that some test is missing :)

**Comment by alexcrichton** [comment]
Closing due to inactivity, but please feel free to resubmit with a rebase!

**Pull request closed without merging**